### PR TITLE
Make thermaltime sklearn compatible

### LINKED ISSUE #151: Call pyphenology thermaltime model from within pycaret
We want to compare ML models to physical models.
Pyphenology has a nice collection of physical models that roughly follows sklearn api
Would be nice if we could make them fully sklearn compliant so we can call them from within pycaret.

### PR DESCRIPTION
**Pull request by SarahAlidoost**
closes https://github.com/phenology/springtime/issues/151

### CODE REVIEW

File: pyPhenology/models/sklearn_thermaltime.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        # TODO check ranges 't1': (-67, 298), 'T': (-25, 25), 'F': (0, 1000)
+
+    def fit(
+        self,
+        X,
+        y,
+    ):
+        # Check that X and y have correct shape
+        X, y = check_X_y(X, y)
+        # raise Exception(X)
```
Comment (Peter9192) [inline pyPhenology/models/sklearn_thermaltime.py]:
```suggestion
        
```

File: pyPhenology/models/sklearn_thermaltime.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        self.X_ = X
+        self.y_ = y
+
+        # Define temperature, doy_series
+        self.model_ = ThermalTime()
+        predictors = pd.DataFrame(
+            {
+                "year": len(X.flatten()) * [2000],
```
Comment (Peter9192) [inline pyPhenology/models/sklearn_thermaltime.py]: This seems like a very particular way to derive a list of years from incoming data. Can you clarify how/why this works?

Comment (Peter9192) [reply pyPhenology/models/sklearn_thermaltime.py]: Also could you get `len(X.flatten())` by accessing `X.size` instead?

File: pyPhenology/models/sklearn_thermaltime.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                "site_id": np.repeat(np.arange(len(X)), X.shape[1]),
```
Comment (Peter9192) [inline pyPhenology/models/sklearn_thermaltime.py:58]: Do I understand correctly that your pretending every row is a unique site as a workaround for the expected structure by pyphenology?

File: pyPhenology/models/sklearn_thermaltime.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                "doy": list(range(X.shape[1])) * X.shape[0],
```
Comment (Peter9192) [inline pyPhenology/models/sklearn_thermaltime.py:59]: So here you're assuming that all columns of the incoming data correspondond to a DOY, starting at one and ending at `n`? (where `n` is the number of columns)? Would be good to document that this is the expected structure for this wrapper.

File: pyPhenology/models/sklearn_thermaltime.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    ):
+        # Check that X and y have correct shape
+        X, y = check_X_y(X, y)
+        # raise Exception(X)
+        self.X_ = X
+        self.y_ = y
+
+        # Define temperature, doy_series
+        self.model_ = ThermalTime()
+        predictors = pd.DataFrame(
```
Comment (Peter9192) [inline pyPhenology/models/sklearn_thermaltime.py:55]:
```suggestion
        # Convert incoming data to expected structure as documented here
        # https://pyphenology.readthedocs.io/en/master/data_structures.html
        predictors = pd.DataFrame(
```

File: test/test_sklearn_thermaltime.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        observed_doy = observations_test.doy.values
+        observations_train = observations[10:]
+        model.fit(observations_train, predictors)
+        predicted_doy = model.predict(observations_test, predictors)
+        rmse_phenology = np.sqrt(np.mean((predicted_doy - observed_doy) ** 2))
+
+        # Prepare data to train the model using pycaret
+        doy_array, temperature_array, _ = mu.misc.temperature_only_data_prep(
+            observations, predictors, for_prediction=False
+        )
```
Comment (Peter9192) [inline test/test_sklearn_thermaltime.py:71]: So here we're using pyphenology's internals to tranform the data to the "sklearn" format. Presumably pyphenology also uses this transformation internally. So that means, in our wrapper, we're going from sklearn compatible to pyphenology compatible, and then inside pyphenology we go back to sklearn compatible. 

I wonder if it isn't possible to hook into the existing pyphenology code only after this transformation - that would save us having to transform back and forth between these two formats alltogether.

**Review by Peter9192 — CHANGES_REQUESTED** [review]
Nice work @SarahAlidoost, I like how you test by comparing "raw" vs "wrapped" pyphenology. It's also a quite elegant and simple wrapper - we could consider shipping it on it's own and simply importing pyphenology. 

In either case,  a bit of explanation of how to use this wrapper on top of the readme or so would be very helpful. 

My main question though is about the transformations back and forth, and generalization to other models. It looks like were going through a bit of a hassle to get to the data format pyphenology exposes, but then internally pyphenology more or less tranforms back to where we were coming from. Sounds like with a bit of effort, we could eliminate all these transforms alltogether - though we might have to come up with a standardized column naming scheme. 

My other question is about generalizing this to other pyphenology models. Opened #1 to discuss that further.

File: pyPhenology/models/sklearn_thermaltime.py (hunk 2 above)
Comment (SarahAlidoost) [reply pyPhenology/models/sklearn_thermaltime.py]: More comments are added. Also, `len(X.flatten())` is replaced with `X.size`

File: pyPhenology/models/sklearn_thermaltime.py (hunk 3 above)
Comment (SarahAlidoost) [reply pyPhenology/models/sklearn_thermaltime.py:58]: More comments are added.

File: pyPhenology/models/sklearn_thermaltime.py (hunk 4 above)
Comment (SarahAlidoost) [reply pyPhenology/models/sklearn_thermaltime.py:59]: More comments are added

File: test/test_sklearn_thermaltime.py (hunk 6 above)
Comment (SarahAlidoost) [reply test/test_sklearn_thermaltime.py:71]: yes, good point. I submitted issue #3

**Comment by SarahAlidoost** [comment]
> Nice work @SarahAlidoost, I like how you test by comparing "raw" vs "wrapped" pyphenology. It's also a quite elegant and simple wrapper - we could consider shipping it on it's own and simply importing pyphenology.
> 
> In either case, a bit of explanation of how to use this wrapper on top of the readme or so would be very helpful.
> 
> My main question though is about the transformations back and forth, and generalization to other models. It looks like were going through a bit of a hassle to get to the data format pyphenology exposes, but then internally pyphenology more or less tranforms back to where we were coming from. Sounds like with a bit of effort, we could eliminate all these transforms alltogether - though we might have to come up with a standardized column naming scheme.
> 
> My other question is about generalizing this to other pyphenology models. Opened #1 to discuss that further.

thanks for the helpful review. I addressed your comments. Please have another look. Meantime, I found how to set `seed` in pyPhenology, see my [commit](https://github.com/phenology/pyPhenology/pull/1/commits/06cb19a15a7d901e7d1153763bdc0325b8663c89).

**Review by Peter9192 — APPROVED** [review]
Thanks for addressing the comments! Let's continue the great work in #2 and or #3 👍
